<div id="ABC">
<div style="padding:5px; border:1px solid #000000; background-color:#C0C0C0; font-size:300%; font-weight:400; color: #000000; width:100%;">
Thomas Hollis' Journal
<div style="padding:5px; margin-top:20px; margin-bottom:10px; background-color:#C0C0C0; font-size:30%; font-weight:200; color: #000000; ">
(Thomas Hollis' personal Journal)
</div>
</div>


 __TOC__


= 0. Course preparations =
This section outlines the work done on all the courses that have to do with course preparation. 

== 0.1 [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-Wiki_editing Wiki Editing] ==
; Objective
The main objective is to gain an in depth understanding of how to navigate and edit the student wiki as well as comprehend what will be expected of me throughout the course. 

<div class="time-estimate">
Time estimated: 2 h; taken: 6 h; date started: 2018-09-14; date completed: 2018-09-15
</div>

; Activities
I accessed and created my main StudentWiki user page, journal page and insights page.  This was not too challenging as I have been contributing to Wikipedia's public body of information for a few years now and I am used to the WikiMarkup syntax. As an avid fan of the open-source community, I also created and committed my current progress to my own personal [https://github.com/PsiPhiTheta "Bioinformatics" Github repository] for the rest of the world to use. What was the most time consuming was to explore the entire wiki to see everything that this course has to offer, including all the activities that occurred in previous years. I must confess I did get a little carried away in my exploration and the boundary between "fun" and "work" is already starting to blur. However, I do believe that this is a good sign and I am very excited for the course ahead.

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-Journal Journal Writing].

== 0.2 [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-Journal Journal Writing] ==
; Objective
The main objective is to start off my Journal with a strong baseline on which I can build the rest of my discoveries in this course.

<div class="time-estimate">
Time estimated: 1 h; taken: 1.5 h; date started: 2018-09-15; date completed: 2018-09-15
</div>

; Activities
I started writing my journal, including this precise section. I also edited my main [[User:Thomas_Hollis | user page]] and added some information to help my classmates identify who I am and what my background is. This will undoubtably be useful for collaboration in the future, especially given my computer science background with very little biology knowledge. 

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/ABC-Plagiarism Plagiarism].

== 0.3 [http://steipe.biochemistry.utoronto.ca/abc/index.php/ABC-Plagiarism Plagiarism] ==
; Objective
The main objective is to revise concepts related to Plagiarism. Collaboration can all to easily turn into plagiarism in academic institutions and the only way to stamp it out is through education. 

<div class="time-estimate">
Time estimated: 20 min; taken: 1 h; date started: 2018-09-15; date completed: 2018-09-15
</div>

; Activities
Read in detail the wiki page and the websites referred to in the main body. Checked all existing content to make sure it is fully cited. Added a citation for Google Deep Dream algorithms on my main User: page. Added the following practice APA citations:
* a procedure in the methods section of a journal article, as you would cite it in a technical report:
Gruber, H., Holzer, M., & Ruepp, O. (2016). Sorting the Slow Way: An Analysis of Perversely Awful Randomized Sorting Algorithms. Lecture Notes in Computer Science Fun with Algorithms, 4475, 183-197. doi:10.1007/978-3-540-72914-3_17
* a piece of code you found in a StackOverflow article, as you would put it as a comment into computer code:
<code> # BioGeek. (2015, April 24). Are there any worse sorting algorithms than Bogosort (a.k.a Monkey Sort)? Retrieved September 16, 2018, from https://stackoverflow.com/questions/2609857/are-there-any-worse-sorting-algorithms-than-bogosort-a-k-a-monkey-sort </code>
* some contents from a classmate's journal that you incorporate into your own journal:
Moskal, N. (2013) Open Project Concept. Retrieved September 16, 2018, from http://steipe.biochemistry.utoronto.ca/abc/students/index.php/User:Natalia_Moskal/Open_Project

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/ABC-Insights Insights].

== 0.4 [http://steipe.biochemistry.utoronto.ca/abc/index.php/ABC-Insights Insights] ==
; Objective
The main objective is to learn how to create and maintain the "insights!" page in BCH441. 
<div class="time-estimate">
Time estimated: 0.5 h; taken: 1 h; date started: 2018-09-16; date completed: 2018-09-16
</div>

; Activities
Edited my [[User:Thomas_Hollis/insights! | insights]] page.

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-Netiquette Netiquette].

== 0.5 [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-Netiquette Netiquette] ==
; Objective
The main objective is to learn how to interact with other users on this StudentWiki network and in the mailing list. 
<div class="time-estimate">
Time estimated: 0.5 h; taken: 0.5 h; date started: 2018-09-16; date completed: 2018-09-16
</div>

; Activities
Read the [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-Netiquette Netiquette] page carefully and came up with an insight for future me.

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-Backups Backups].

== 0.6 [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-Backups Backups] ==
; Objective
The main objective is to refresh existing knowledge on how to securely backup data to prevent accidental data loss. 
<div class="time-estimate">
Time estimated: 0.5 h; taken: 0.5 h; date started: 2018-09-16; date completed: 2018-09-16
</div>

; Activities
Read the [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-Backups Backups] page carefully and verified the effectiveness of my current backup solution. I currently store all of my personal files (excluding application data) in iCloud drive, which is continually backed up on Apple servers. So when is the last time I made a backup? Approximately three seconds ago when I saved my most recent file. Apple servers themselves are geographically distributed, redundant and backed up (often using RAID or similar technology). In addition, as I do not trust anyone else but myself with my own data, I also make differential local backups of my entire drive on the 1st of each month. This is done on two separate external hard drives (Time Machine really is a good utility). This allows me to have an on-site backup and off-site backup which protects me in the case that there is a fire and both my laptop and my hard drive get destroyed. Indeed the risk of this is non-negligeable as there are 24 000 house fires per year in Canada <ref>SGI Canada. (2018). Fire prevention and safety. Retrieved September 16, 2018, from https://www.sgicanada.ca/news?title=fire-prevention-and-safety</ref> for a total of 14 million houses <ref>Statistics Canada. (2016). Canada 2016 Census. Retrieved September 16, 2018, from https://www12.statcan.gc.ca/census-recensement/2016/as-sa/98-200-x/2016005/98-200-x2016005-eng.cfm</ref>. The probability of fire is therefore around 0.17% per year, roughly the same as the SSD failure rate listed in the [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-Backups Backups] page. 

As a quick sanity check I created and deleted a test file (containing the data: "ATGC") that I successfully recovered, as shown aside. [[File:RecoveredFile.png|200px|thumb|right|Recovered data from deleted test file]]

; Conclusion and Outlook
While friends may laugh at my backup strategy and fail to understand why I get so excited on the first day of each month, I will persist in backing up my data reliably. The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-Cargo_Cult Cargo Cult].

== 0.7 [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-Cargo_Cult Cargo Cult] ==
; Objective
The main objective is to understand how to better identify examples of Cargo Cult science.
<div class="time-estimate">
Time estimated: 0.5 h; taken: 1 h; date started: 2018-09-16; date completed: 2018-09-16
</div>

; Activities
Read the 1974 {{WP|Richard Feynman}} [http://calteches.library.caltech.edu/51/2/CargoCult.htm Caltech Commencement address]. Contributed to the [http://steipe.biochemistry.utoronto.ca/abc/students/index.php/Cargo_Cult_Science Cargo Cult Science] and [http://steipe.biochemistry.utoronto.ca/abc/students/index.php/Not_(quite)_Cargo_Cult_Science Not (quite) Cargo Cult Science] pages by adding the example of [http://steipe.biochemistry.utoronto.ca/abc/students/index.php/Cargo_Cult_Science#SCI_gen SCIgen] and commenting on the [http://steipe.biochemistry.utoronto.ca/abc/students/index.php/Cargo_Cult_Science#SpongeBob_and_his_bubble_dance Spongebob Cargo Cult] example.

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-Technical_questions Technical Questions].

== 0.8 [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-Technical_questions Technical questions] ==
; Objective
The main objectives are to understand how to ask technical questions and to learn the concept of Minimal Working Examples (MWEs).

<div class="time-estimate">
Time estimated: 0.5 h; taken: 0.5 h; date started: 2018-09-16; date completed: 2018-09-16
</div>

; Activities
Read the three recommended documents and bookmarked them for later use. 

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/BIN-Info_sources Information Sources].

== 0.9 [http://steipe.biochemistry.utoronto.ca/abc/index.php/BIN-Info_sources Information Sources] ==
; Objective
The main objectives are to build a database of information sources that will be of invaluable help for subsequent modules and to learn to navigate these resources.

<div class="time-estimate">
Time estimated: 0.5 h; taken 1 h; date started: 2018-09-16; date completed: 2018-09-16
</div>

; Activities
Bookmarked all useful resources and spend some time on each to create an account and/or learn to navigate the website.

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-Biocomputing_setup Biocomputing Setup].

== 0.10 [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-Biocomputing_setup Biocomputing Setup] ==
; Objective
The main objectives are to set up my computer with the required IDEs, languages and utilities needed for bioinformatics. 

<div class="time-estimate">
Time estimated: 1 h; taken: 10 min; date started: 2018-09-16; date completed: 2018-09-16
</div>

; Activities
After reading the entire recommended setup, I made a total of 0 changes to my setup. Indeed, I already had installed all the recommended tools, already structured my filesystem well and of course all my file extensions are always displayed by default. Yay! 

I very much like that Prof. Steipe is on my side in the macOS vs. Windows war... take that [https://en.wikipedia.org/wiki/Bill_Gates Bill]. 

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Installation Installation].

= 1. R Programming =
This section outlines the work done on all the courses related to learning the R programming language. 

== 1.1 [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Installation Installation] ==
; Objective
The main objectives are to remind myself on how to install and run packages in R. 

<div class="time-estimate">
Time estimated: 1 h; taken: 0.5 h; date started: 2018-09-16; date completed: 2018-09-16
</div>

; Activities
Indeed as I have had prior experience using R and RStudio this module was more of a refresher. Nonetheless, I found it to be a useful review - especially the script that checks if a package is already installed before installing it. I also executed all the code suggested and had a good play around with the new packages discussed. 

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Setup Setup].

== 1.2 [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Setup Setup] ==
; Objective
The main objectives are to check that my R environment is setup as recommended for bioinformatics.

<div class="time-estimate">
Time estimated: 0.5 h; taken: 1 h; date started: 2018-09-16; date completed: 2018-09-16
</div>

; Activities
Indeed as I have used RStudio in the past only for machine learning and statistics, it was not optimally configured for Bioinformatics so I slightly tweaked my setup. I also learnt all about connecting RStudio to GitHub and did all the suggested exercises.

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Console Console].

== 1.3 [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Console Console] ==
; Objective
The objective here was to learn the difference between executing entire scripts and executing commands individually from the console. 

<div class="time-estimate">
Time estimated: 0.5 h; taken: 0.1 h; date started: 2018-09-17; date completed: 2018-09-17
</div>

; Activities
This was already known to me but reading the full wiki page is always good to serve as a refresher. 

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Help Help].

== 1.4 [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Help Help] ==
; Objective
The main objective here is to gain familiarity with all the different sources of help for R programming outside what has already been mentioned. 

<div class="time-estimate">
Time estimated: 0.5 h; taken: ?? h; date started: 2018-09-17; date completed: 2018-09-17
</div>

; Activities
Read all the links recommended and tried out the code excerpts.

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Syntax_basics Syntax basics].

== 1.5 [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Syntax_basics Syntax basics] ==
; Objective
The objective is to refresh on some R basics through examples.

<div class="time-estimate">
Time estimated: 0.5 h; taken: 0.5 h; date started: 2018-09-18; date completed: 2018-09-18
</div>

; Activities
Running all the suggested code was a good opportunity for me to check that I am building my knowledge on good foundations. I was also able to make sure I don't have any bad R-programming habits to shake off. One particular thing I brushed up on was the difference between the "<-" and "=" in R. A really good explanation was read [https://renkun.me/2014/01/28/difference-between-assignment-operators-in-r/ here]. I also found out in this week's lecture that you never need to explicitly define parameters that have the "=" sign in R documentation. 

<div class="toccolours mw-collapsible mw-collapsed" style="width:800px">
The code I wrote can be found here:
<div class="mw-collapsible-content">
<source lang="r">
# Thomas Hollis (BCH441, University of Toronto) -v2.1

numbers <- c(16, 20, 3, 5, 9)
numbers

lastNum <- tail(numbers, 1)    
lastNum

if (lastNum < 6 | lastNum > 10) {
  (bool1 <- 1)
} else {
  (bool1 <- 2)
}

if (lastNum >= 10 & lastNum < 20) {
  (bool2 <- 1)
} else {
  (bool2 <- 2)
}

lastNum <- (lastNum/7) - ((((lastNum/7)*10)%/%1)/10)

lastNum <- lastNum*100
lastNum <- lastNum %/% 1
lastNum <- lastNum ^ (1/3)

if (lastNum == 2) {
  (bool3 <- 1)
} else {
  (bool3 <- 2)
}

# END
</source>

</div>
</div>

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Objects-Vectors Vectors].

== 1.6 [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Objects-Vectors Vectors] ==
; Objective
My objective for this unit is to build on my existing knowledge of R vectorisation. 

<div class="time-estimate">
Time estimated: 1 h; taken: 1 h; date started: 2018-09-18; date completed: 2018-09-18
</div>

; Activities
Ran through all the code suggested and found out something that I completely overlooked in the past: the difference between <code>mode(x)</code> and <code>class(x)</code>. A fascinating extension to the comments from Prof. Steipe can be found [https://stats.stackexchange.com/questions/3212/mode-class-and-type-of-r-objects here].

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Objects-Data_frames Data Frames].

== 1.7 [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Objects-Data_frames Data Frames] ==
; Objective
My main goal in this section is to remind myself what an R data frame is and how it operates.

<div class="time-estimate">
Time estimated: 1 h; taken: 1 h; date started: 2018-09-18; date completed: 2018-09-19
</div>

; Activities
I ran through the code and spend a significant amount of time getting lost in rabbit wholes about how R ''actually'' does things... (see more on this in my [http://steipe.biochemistry.utoronto.ca/abc/students/index.php/User:Thomas_Hollis/insights!#Insight_.236_-_Why_you_should_never_trust_someone_who_says_.22Trust_Me.22 insights!] page).

Oh and I actually wrote some code too...

<div class="toccolours mw-collapsible mw-collapsed" style="width:800px">
One line but it counts anyway right?
<div class="mw-collapsible-content">
<source lang="r">
# Thomas Hollis (BCH441, University of Toronto) -v2.1
rownames(plasmidData)[4]<-"pMAL-p5x"
# END
</source>
</div>
</div>

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Objects-Lists Lists].

== 1.8 [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Objects-Lists Lists] ==
; Objective
My objective here is to understand some concepts about lists that I may have overlooked.

<div class="time-estimate">
Time estimated: 0.5 h; taken: 1 h; date started: 2018-09-18; date completed: 2018-09-19
</div>

; Activities
This unit was actually very interesting as it went over in detail an aspect of R that I had always overlooked: the detail of lists. I learn lots of new R tricks! 
<div class="toccolours mw-collapsible mw-collapsed" style="width:800px">
Here is the script I wrote:
<div class="mw-collapsible-content">
<source lang="r">
# Thomas Hollis (BCH441, University of Toronto) -v2.1

pACYC184 <- list(size="4245", marker="Tet, Cam", ori="pUC19")
pUC19 <- list(size=2686, marker="ampicillin", ori="ColE1", accession="L01397", BanI=c(235, 408, 550, 1647) )
plasmidDB <- list()
plasmidDB[["pUC19"]] <- pUC19
plasmidDB[["pACYC184"]] <- pACYC184
pBR322 <- list(size=4361, marker="Amp, Tet", ori="ColE1")
plasmidDB[["pBR322"]] <- pBR322

plasmidDB[["pACYC184"]]

#Now we must retrieve all sizes from the list and find the smallest
lapply(plasmidDB, function(x) {return(x$ori)})
x <- unlist(lapply(plasmidDB, function(x) {return(x$size)}))
min(x)

# END
</source>
</div>
</div>

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Subsetting Subsetting].

== 1.9 [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Subsetting Subsetting] ==
; Objective
The main goal for me here is to learn the details of subsetting in R which I may not already be aware of.

<div class="time-estimate">
Time estimated: 1 h; taken: 2 h; date started: 2018-09-19; date completed: 2018-09-19
</div>

; Activities
I was surprised to learn a lot of new things about subsetting when completing all the required exercises.  For example, it was interesting to find out the difference between <code>order(x)</code> and <code>sort.list(x)</code> simply by reading R documentation! I absolutely love how organised the R documentation is - makes coding in R extremely smooth! Lots of new things covered here. I must make sure to try to practice what I have learnt soon to help engrain it into long term memory.

''Reminder from past-thomas to future-thomas: ''

''Do not write code where you use the magic numbers: name your rows and columns instead!''

<div class="toccolours mw-collapsible mw-collapsed" style="width:800px">
Here is the code I wrote for this unit:
<div class="mw-collapsible-content">
<source lang="r">
# Thomas Hollis (BCH441, University of Toronto) -v2.1

LPSdat[1:10, 1:2] #first ten rows of first two columns
LPSdat[10:1, 1:2] #first ten rows backward of first two columns
LPSdat[sample(1:10), 1:2] #first ten rows in random order of first two columns
LPSdat[order(LPSdat[1:10,2]),1:2] #first ten rows of first two columns ordered by value in the second column ascending
LPSdat$Mo.LPS[1:10] #first ten rows of colum named Mo.LPS
LPSdat[1:10, c("Mo.LPS","Mo.ctrl")] #first ten rows of column Mo.LPS and Mo.ctrl
LPSdat$genes[nchar(LPSdat$genes)==3] #all genes with gene name over three characters long
LPSdat[grep("Il",LPSdat$genes), 1:2] #first two columns of all genes that have the string "Il" inside their name
LPSdat[(LPSdat$B.LPS - LPSdat$B.ctrl) > 2,] #all genes for which B-cells are stimulated by LPS by more than 2 log units

# END
</source>
</div>
</div>

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Control_structures Control Structures].

== 1.10 [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Control_structures Control Structures] ==
; Objective
I think that my main task for this section will be to learn about the <code>seq_along()</code> function as the other material I believe will only serve as a reminder.

<div class="time-estimate">
Time estimated: 1 h; taken: 1 h; date started: 2018-09-19; date completed: 2018-09-19
</div>

; Activities
Updated my knowledge on control structures - especially now I know that using <code>seq_along()</code> prevents unexpected behaviour for <code>null</code> sequences.
<div class="toccolours mw-collapsible mw-collapsed" style="width:800px">
Code developed:
<div class="mw-collapsible-content">
<source lang="r">
# Thomas Hollis (BCH441, University of Toronto) -v2.1

cnt <- 3 #numeric count
txt <- as.character(cnt) #character count
while (cnt > -1) { #break condition
    txt <- c(txt, cnt) #append new character to count
    cnt = cnt - 1 #decrement
}
(txt <- c(txt, "Lift Off!")) #append final string & print

# END
</source>
</div>
</div>

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Functions Functions].

== 1.11 [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Functions Functions] ==
; Objective
This unit will mostly be practice in consolidating my skills at writing functions in R.

<div class="time-estimate">
Time estimated: 1 h; taken: 1.5 h; date started: 2018-09-19; date completed: 2018-09-19
</div>

; Activities
My main activity was to have my mind blown: inputs to R functions can be abbreviated up to their smallest unique form! I never knew that! Oh and I had an extended practice by modifying to the extreme some of the functions provided. 
<div class="toccolours mw-collapsible mw-collapsed" style="width:800px">
Code developed:
<div class="mw-collapsible-content">
<source lang="r">
# Thomas Hollis (BCH441, University of Toronto) -v2.1 

countDown <- function(n) {
    cnt <- n #numeric count
    txt <- as.character(cnt) #character count
    while (cnt > 0) { #break condition
        cnt = cnt - 1 #decrement
        txt <- c(txt, cnt) #append new character to count
    }
    return (txt <- c(txt, "Lift Off!")) #append final string & print
}

myLifeDays <- function(birthday, reverse) {
    if (missing(birthday)) {
        if (missing(reverse)) { #if no arguments are provided
            print ("Enter your birthday as a string in \"YYYY-MM-DD\" format and/or the days as an integer.")
            return()
        }
    }

    if (missing(reverse)) { #if only the birthday argument is provided
        bd <- strptime(birthday, "%Y-%m-%d") # convert string to time
        now <- format(Sys.time(), "%Y-%m-%d") # convert "now" to time
        diff <- round(as.numeric(difftime(now, bd, unit="days")))
        print(sprintf("This date was %d days ago.", diff))
    } else if (missing(birthday)) { #if only the reverse argument is provided
        now <- format(Sys.time(), "%Y-%m-%d")
        print(as.Date(now) + reverse)
    } else { #if both arguments are provided
        bd <- strptime(birthday, "%Y-%m-%d") # convert string to time
        now <- format(Sys.time(), "%Y-%m-%d") # convert "now" to time
        diff <- round(as.numeric(difftime(now, bd, unit="days")))
        print(sprintf("This date was %d days ago.", diff))

        now <- format(Sys.time(), "%Y-%m-%d")
        print(as.Date(now) + reverse)
    }
}

myLifeDays("1932-09-25", 3)
myLifeDays("1932-09-25")
myLifeDays(,3)
myLifeDays(r=3,b="1970-01-01")

countDown(10)

seq(-5, 3) #sequence of integers from -5 to 3
seq(from = -2, to = 2, by = (1/3)) #sequence of numbers from -2 to 2 in intervals of 1/3
seq(length.out=30, to=100, from=1) #sequence of 30 numbers between 1 and 100. Pass the arguments in the following order: length.out, to, from.

# END
</source>
</div>
</div>

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Plotting Plotting].

== 1.12 [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Plotting Plotting] ==
; Objective
Here I would like to learn new methods of plotting and displaying data in R as it is a crucial skill.

<div class="time-estimate">
Time estimated: 1 h; taken: 1.5 h; date started: 2018-09-19; date completed: 2018-09-20
</div>

; Activities
Sat in mesmerisation as all the nice colours of R plots flowed from my screen into my eyes. Then saved a local copy of the amazing tutorial I just completed because that was absolutely fantastic! Future-me is going to love this.

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Coding_style Coding Style].

== 1.13 [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Coding_style Coding Style] ==
; Objective
Learn a consistent coding style to use throughout the Wiki! 

<div class="time-estimate">
Time estimated: 0.5 h; taken: 1 h; date started: 2018-09-20; date completed: 2018-09-20
</div>

; Activities
Well let's just say I may have re-editted all the code I have written till now to make it conform to the standards expected...

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Introduction Introduction to R].

== 1.14 [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Introduction Introduction to R] ==
; Objective

<div class="time-estimate">
Time estimated: 0.5 h; taken: 0.5 h; date started: 2018-09-24; date completed: 2018-09-25
</div>

; Activities
Git cloned the ABC-units repository and ran the sample task to ensure R studio is set up correctly. All good. 

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-Cell_cycle Cell Cycle].

== 1.15 [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Unit_testing Testing R Code] ==
; Objective
My main objective here is to follow a rigorous protocol for testing my future R code. 

<div class="time-estimate">
Time estimated: 0.5 h; taken: 0.5 h; date started: 2018-10-05; date completed: 2018-10-05
</div>

; Activities
[[File:FloatingPointIssues.jpg|200px|thumb|right|Floating Point Issues]]
I followed the learning unit page thoroughly and ran through the full R script. Floating point errors always made me very very sad (see aside). But what makes me even more sad is that <code> (0.1 + 0.7) == 0.8 </code> returns <code> FALSE </code>! Absolute abomination! To make things worse the reason behind this is just as tragic: [https://cran.r-project.org/doc/FAQ/R-FAQ.html#How-can-I-clean-up-my-workspace_003f FAQ section 7.31].

<div class="toccolours mw-collapsible mw-collapsed" style="width:800px">
Oh and I also wrote some code
<div class="mw-collapsible-content">
<source lang="r">
# Thomas Hollis (BCH441, University of Toronto) -v2.1

calcGC <- function(inputSeq) {
  countBefore <- nchar(inputSeq)
  inputSeq <- gsub("G", "", inputSeq)
  inputSeq <- gsub("C", "", inputSeq)
  countAfter <- nchar(inputSeq)
  count = (countBefore - countAfter)/countBefore
  return(count)
}

expect_equal(calcGC("AATT"), 0)
expect_equal(calcGC("ATGC"), 0.5)
expect_equal(calcGC("AC"), 0.5)
expect_equal(calcGC("CGCG"), 1)

# END
</source>
</div>
</div>

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-STA-Probability Probability].

== 1.16 [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-FASTA FASTA] ==
; Objective
My main objective here is to understand the structure of FASTA files. 

<div class="time-estimate">
Time estimated: 0.5 h; taken: 0.5 h; date started: 2018-10-23; date completed: 2018-10-24
</div>

; Activities
I went through the entire programming tutorial. 

<div class="toccolours mw-collapsible mw-collapsed" style="width:800px">
Sadly there was no code for me to write here...
<div class="mw-collapsible-content">
<source lang="r">
# You were just too curious to see what was in here weren't you... well, like I said, nothing. (But since you're here to avoid wasting your time here is a little present: https://en.wikipedia.org/wiki/Belphegor%27s_prime)
</source>
</div>
</div>

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Biostrings Biostrings].


== 1.17 [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Biostrings Biostrings] ==
; Objective
My main objective here is to gain familiarity with the Biostrings R package. 

<div class="time-estimate">
Time estimated: 0.5 h; taken: 0.5 h; date started: 2018-10-24; date completed: 2018-10-24
</div>

; Activities
The usual <code>init()</code> and run through the tutorial goodness.

<div class="toccolours mw-collapsible mw-collapsed" style="width:800px">
Sadly there was no code for me to write here either...
<div class="mw-collapsible-content">
<source lang="r">
# Fool me once, shame on you. Fool me twice? No presents for you.
</source>
</div>
</div>

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/BIN-ALI-Similarity Measuring Sequence Similarity].

== 1.18 Extra learning units ==
; Objective
The main objective is to read through extra programming modules that are not necessary to complete the course. 

<div class="time-estimate">
Time estimated: 1 h; taken: 1 h; date started: 2018-09-25; date completed: 2018-09-25
</div>

; Activities
I had a good read through of the [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Debugging Debugging] and [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Literate_programming Literate Programming] courses! Sadly, the unit that appealed most to me, [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Optimization Optimization], is still a stub.

; Conclusion and Outlook
Although this was not required, it was definitely worthwhile and good fun.

= 2. Fundamentals =
This section outlines the work done on all the courses that have to do with bioinformatics fundamentals. 

== 2.1 [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-Cell_cycle Cell Cycle] ==
; Objective
Understand the fundamentals behind the cell cycle, in particular for the case of yeast and the role of the Mbp1 protein.

<div class="time-estimate">
Time estimated: 1 h; taken: 2 h; date started: 2018-09-25; date completed: 2018-09-25
</div>

; Activities
Very intensive reading followed by more intensive reading. All the sources provided (as well as a few visits to Wikipedia) where explored. Much of the content was not understood so I am going to come back to this. Wow, that's a lot of biology I need to catch up on! Thanks to [http://steipe.biochemistry.utoronto.ca/abc/students/index.php/User:Alana_Man/BCH441_Journal Alana Man] for making me realise how much biology knowledge I need to work to catch up on! Her journal is a good sanity check when I find a certain unit challenging. 

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-CSC-Data_models Data Models].

== 2.2 [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-CSC-Data_models Data Models] ==
; Objective
Learn how to sketch data models.

<div class="time-estimate">
Time estimated: 1 h; taken: 2 h; date started: 2018-09-30; date completed: 2018-10-04
</div>

; Activities
I wrote my first bioinformatics sketch! Also I was the first one in the class to do so, so I am happy with the pace at which I am currently working. I don't think my sketch is very good but I look forward to all the feedback from the mailing list! Here is the first version:

<div class="toccolours mw-collapsible mw-collapsed" style="width:800px">
I am so ashamed of it that I have hidden it inside this collapsible box.
<div class="mw-collapsible-content">
[[File:TomProteinDB.png]]
</div>
</div>

Indeed, from all the feedback I received I found out that I had completely missed a critical part of the unit which was a link to a pdf document about data models. Since I had totally not seen this I was a bit confused as I guessed my sketch out of the blue. One humiliating lecture painfully revealed how much I have missed. This was however a surprisingly a great learning opportunity to ensure I am more careful in the future. I redid this unit and wrote an amended version of my initial abomination:

[[File:TomProteinDB2.png]]

I feel like the toughest part here was understanding the biology behind it rather than actually building the model... I had to google A LOT of things. I may revisit this to improve the model at a later date.

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/BIN-MYSPE My Species].

== 2.3 [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-CSC-Software_development Software Development] == 
; Objective
Learn the formal theory behind software development, something that I have taken for granted for all my years as a software engineer. 

<div class="time-estimate">
Time estimated: 1 h; taken: 1 h; date started: 2018-10-05; date completed: 2018-10-05
</div>

; Activities
Read through in detail the contents of the learning unit page.

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-CSC-Test_driven_development Test Driven Development].

== 2.4 [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-CSC-Test_driven_development Test Driven Development] == 
; Objective
Learn what test driven development is and how to leverage it in bioinformatics.

<div class="time-estimate">
Time estimated: 1 h; taken: 0.5 h; date started: 2018-10-05; date completed: 2018-10-05
</div>

; Activities
Read through in detail the contents of the learning unit page. Indeed I had actually done some test-driven development before but I was simply not aware that this development approach is called like that. I find it to be often useful but tend to overlook this approach for smaller projects. I can totally understand why it would be critical for bioinformatics and look forward to returning to this development style in subsequent work.

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Unit_testing Testing R Code].

== 2.5 [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-MAT-Graphs_and_networks Graphs & Networks] (submitted: 6 marks)== 
; Objective
The objective in this unit is to understand how relationships between genes are presented in various graphs & networks. Much time was dedicated here as I plan to maximise my performance in the quiz that will take on this unit. 

<div class="time-estimate">
Time estimated: 6 h; taken: 6 h; date started: 2018-10-13; date completed: 2018-10-23
</div>

; Activities
I will take the in class quiz for this learning unit (yay, I got 100%)! I also attended a quiz before on the NCBI and EBI, just to get a feeling of how they are. 

I read through the notes on graphs and networks and went through the R script tutorial, which I particularly enjoyed. I also had a gander at the [https://www.ncbi.nlm.nih.gov/pubmed/21527005 pubmed paper] on analysing biological networks using graph theory. A very interesting read even if some of the details went over my head.

- The 8 main categories of graphs covered (in lectures and in my external readings) were: DAGs (directed acyclic graphs), cyclic graphs (like DAGs but with infinite response), trees (top to bottom direction only graph), random graphs, hypergraphs (with sets), dual graphs (a.k.a Voronoi diagrams), De Bruijn graphs (with amino acid sequences), Euler cycles (each edge visited once), Hamilton Cycles (each node visited exactly once).

- The 5 main types of graph set metrics covered were: node number (size), edge number (in-degree, out-degree or simply degree), histogram of degrees, topology (degree centrality leads to high degree nodes in centre of network), state (random, scale-free i.e. no meaningful average node or hierarchical).

- The 5 main topologic metrics covered are: shortest path, centrality, diameter, spanning trees (all the vertices covered with minimum possible number of edges - cannot be disconnected), causality (cannot have time delay units violating principles of causality) 

- The 4 main methods for automatic graph generation were: Erdos-Renyi (random connection), Gilbert, Barabasi-Albert (degree drop as node number increases), Price.

- The 4 main techniques for achieving scale-free graphs were: Erdos-Renyi (as before), copy model (add new nodes by copying a fraction of links of existing nodes), hierarchal model, hyperbolic geometric graphs (nodes connected if distance between them is smaller than a hyperbolic threshold)

- The 2 main algorithms for structure analysis covered were: Dijkstra's algorithm (find shortest path in O(<math>V^2</math>)), Floyd-Warshall algorithm (all the shortest paths in O(<math>V^3</math>))

The main equations to remember here are:

- Scale free state probability (where <math>k</math> is usually between 2 and 3): 

<math>P(k) \propto k^{-\gamma}</math> (blimey, wiki renders LaTeX poorly...)

- Betweenness centrality: 

<math>C_b(w) = \sum_{i\neq j\neq w} \frac{\sigma_{ij}(w)}{\sigma_{ij}}</math> 

; Conclusion and Outlook

Graph theory is so awesome. I was familiar with Dijkstra before (a personal hero of mine) but I had no idea about many metrics discovered today. A true gem of information. I can see how these graphs could be leveraged frequently in bioinformatics! 

The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/BIN-Databases Databases].

== 2.6 [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-Genetic_code Genetic Code] ==
; Objective
The main objective here was to learn how DNA leads to protein synthesis through the use of mRNA and codons.

<div class="time-estimate">
Time estimated: 1 h; taken: 1 h; date started: 2018-10-06; date completed: 2018-10-06
</div>

; Activities
Thankfully I read a lot of the supplementary materials including much of the wikipedia content. I think without this, the coding exercise would have been very confusing. I find all this stuff so cool. My world of computer science crashing into my genome, it's a beautiful sight. 

I was surprised to find a lot of M codons (start codons) and did some digging as to why this is the case. In fact I found on [https://biology.stackexchange.com/questions/46427/multiple-start-and-stop-codons-in-mrna-and-pre-mrna the Biology StackExchange] that in about 5% of genes the first AUG (start codon) is skipped and translation starts at one of the other AUG sequences (although this is still poorly understood). 

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/BIN-Sequence Sequence]

== 2.7 Extra learning units ==
; Objective
The main objective is to read through extra fundamental modules that are not necessary to complete the course to make sure I am fully equipped. 

<div class="time-estimate">
Time estimated: 0.5 h; taken: 0.5 h; date started: 2018-10-01; date completed: 2018-10-05
</div>

; Activities
So far most the extra units I have checked out ([http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-CSC-Systems_models Systems Models], [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-CSC-Integration_testing Integration testing], [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-STA-Bayes_theorem Bayes Theorem], [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-STA-Enrichment Enrichment] and [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-STA-Hypothesis_testing Hypothesis Testing]) are all still stubs. However, [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-CSC-SPN Structured Process Notation] was a very interesting read!

; Conclusion and Outlook
Although this was not required, it was definitely worthwhile!

= 3. Bioinformatics =
This section outlines the work done on all the courses that have to do with bioinformatics fundamentals. 

== 3.1 [http://steipe.biochemistry.utoronto.ca/abc/index.php/BIN-MYSPE My Species] ==
; Objective
Acquire my own species that I will work on for the rest of the course.

<div class="time-estimate">
Time estimated: 1 h; taken: 1 h; date started: 2018-10-05; date completed: 2018-10-05
</div>

; Activities
Read through the learning unit and ran the R code. My species is "Babjeviella inositovora" and it's BiCode is "BABIN". This is awesome, I love it! I even created a [https://en.wikipedia.org/wiki/Babjeviella_inositovora Wikipedia page] for it! Why does El Salvador have a page for it and not the english wiki? According to its [https://www.ncbi.nlm.nih.gov/genome/?term=Babjeviella+inositovora NCBI page], my species has potential applications in biotechnology.

Update: due to an issue in the BABIN genome sequence being incomplete, Prof. Steipe has asked me to abandon it and pick a new one (by incrementing my student number by 1). This was quite sad as I was starting to get attached to BABIN. Oh well, my new species is "Isaria fumosorosea" and it's BiCode is "ISAFU". It already has a [https://en.wikipedia.org/wiki/Isaria_fumosorosea Wikipedia page] and I look forward to adopting it. It looks like a lot of research has already been done on it, as it shows promise of being a good pesticide!

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/BIN-Abstractions Abstractions].

== 3.2 [http://steipe.biochemistry.utoronto.ca/abc/index.php/BIN-Abstractions Abstractions] ==
; Objective
Learn how abstractions are made in the field of bioinformatics.

<div class="time-estimate">
Time estimated: 1 h; taken: 1 h; date started: 2018-10-05; date completed: 2018-10-05
</div>

; Activities
Interestingly enough there are many similarities between the abstractions I am used to making in computer science (such as those made by [https://en.wikipedia.org/wiki/Directed_acyclic_graph Directed Acyclic Graphs]) and those made in bioinformatics. This was my first bioinformatics abstraction attempt:
[[File:TomAbstraction.png]]

Although it is not very complete it came closer than I expected to the actual abstraction that is commonly used: [https://en.wikipedia.org/wiki/Gene_regulatory_network Gene Regulatory Networks].

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-CSC-Software_development Software Development].

== 3.3 [http://steipe.biochemistry.utoronto.ca/abc/index.php/BIN-Sequence Sequence] ==
; Objective
Learn how sequences work in bioinformatics, what types of sequences exist and how to use them in my own work.

<div class="time-estimate">
Time estimated: 1 h; taken:1 h; date started: 2018-10-06; date completed: 2018-10-06
</div>

; Activities
I must confess I have not yet memorised all of the IUPAC one letter amino acid codes. I also was not able to answer all of the [http://steipe.biochemistry.utoronto.ca/abc/index.php/Amino_Acid_Exam_Questions past exam questions] but I believe this might be because of my lack of biological knowledge. The R code was certainly interesting to go through too! 

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Genetic_code_optimality Genetic Code Optimality]

== 3.4 [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-Genetic_code_optimality Genetic Code Optimality] ==
; Objective
The main objective here is to look at alternative genetic codes and evaluate their robustness and optimality with regard to code changes (mutations).

<div class="time-estimate">
Time estimated: 1 h; taken: 1 h; date started: 2018-10-06; date completed: 2018-10-06
</div>

; Activities
Followed the code in depth. Very interesting stuff. A very powerful demonstration of genetic robustness!


<div class="toccolours mw-collapsible mw-collapsed" style="width:800px">
The code I wrote can be found here:
<div class="mw-collapsible-content">
<source lang="r">
# Thomas Hollis (BCH441, University of Toronto) -v2.1

set.seed(pi)
N <- 200
valSGC <- numeric(N)

for (i in 1:N) {
  GC <- swappedGC(GENETIC_CODE)   
  x <- traRev(myAA, GC)         
  x <- randMut(x)                
  x <- traFor(x, GC)             
  valXGC[i] <- evalMut(myAA, x)  
}

hist(valXGC,
     col = "turquoise",
     breaks = 15,
     add = TRUE)

# END
</source>

</div>
</div>

[[File:TomGeneticCodeOptimality.png]]

Interpretation: Both modified alternative genetic codes yielded distributions with significantly more mutations (based on mutation frequency & penalty as defined in the code). Interestingly, the variance of these mutation distributions was also wider for the modified genetic codes (in blue and purple) than for the original (in yellow). This suggests that the original genetic code is particularly good at resisting the impact of mutations. Indeed, even after some changes, these mutations will not necessarily be as expressed in the original genetic code than it would be in the alternative genetic codes. 

; Conclusion and Outlook
The genetic code is remarkably mutation proof. This is because the impact of the mutations (once translated back) on average causes less penalty (tangible impact) than in randomly generated variations of the genetic code. This would be in line with my expectations as I would expect natural selection to leave room for some mutation but not an excessive and unmanageable amount (with my limited knowledge of biology though, I am not making this statement with much confidence or authority...).

The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/BIN-Storing_data Storing data]

== 3.5 [http://steipe.biochemistry.utoronto.ca/abc/index.php/BIN-Storing_data Storing Data] (submitted: 6 marks)==
; Objective
The main goal here is to learn about multiple concepts regarding data storage and ACID databases (for example JSON, mySQL and their R wrappers).

<div class="time-estimate">
Time estimated: 6 h; taken: 6 h; date started: 2018-10-07; date completed: 2018-10-13
</div>

; Activities
My first learning unit to be submitted for credit! I intend on spending a lot of time on this to make sure I get it right. I created the evaluation submission page at: [http://steipe.biochemistry.utoronto.ca/abc/students/index.php/User:Thomas_Hollis/Eval_Storing_Data User:Thomas_Hollis/Eval_Storing_Data]. 

The data formats covered are: '''FASTA''' (simple, commonly used), '''GBFF''' (extensively annotatable), '''BLAST''' (output file parsers heavily worked on for many years), '''PDB''' (apparently impossible to read all information from all PDB files with a single parser).

The data grammars covered are: '''ASN.1''' (abstract syntax notation v1 - used by the NCBI), '''XML''' (extended markup language - widely used online), '''JSON''' (Java Script Object Notation - popular for key:value pairs, more human readable than XML). 

The data model implementations covered are: '''text files in filesystems''' (using for example a folder hierarchy containing JSON files), '''spreadsheets''' (do not scale well, convert to .csv files for R imports), '''R lists and dataframes''' (what will be used mostly in this course), '''relational databases''' (such as MySQL, Maria DB, or Postgres).

I also went through all the R code thoroughly with some extra background reading to make sure everything is in tip top shape. I had to change MYSPE as BABIN had missing information. Information on my new MYSPE can be found [http://steipe.biochemistry.utoronto.ca/abc/students/index.php/User:Thomas_Hollis/Journal#3.1_My_Species here].

Code developed:

<source lang="r">
# Thomas Hollis (BCH441, University of Toronto) -v2.1 
#     - Purpose: Add an entry to a database and output information in a desired format
#     - Bugs & issues: no bugs, no issues, no warnings

# --- SECTION 1: Add entry to database -----------------
#    Write and submit code that adds another philosopher to the datamodel:
#       Immanuel Kant, (1724 - 1804), Enlightenment Philosophy.
#       Works: Critique of Pure Reason (1781), Critique of Judgement (1790)

tempPersonID <- autoincrement(philDB$person)
temp <- data.frame(id = tempPersonID,
                  name = "Immanuel Kant",
                  born = "1724",
                  died = "1804",
                  school = "Enlightenment Philosophy",
                  stringsAsFactors = FALSE)
philDB$person <- rbind(philDB$person, temp)

rm(temp) #I like my code like I like my room: clean and tidy :)

tempBookID1 <- autoincrement(philDB$books)
temp <- data.frame(id = tempBookID1,
                  title = "Critique of Pure Reason",
                  published = "1781",
                  stringsAsFactors = FALSE)
philDB$books <- rbind(philDB$books, temp)

rm(temp) #I like my code like I like my room: clean and tidy :)

tempBookID2 <- autoincrement(philDB$books)
temp <- data.frame(id = tempBookID2,
                   title = "Critique of Judgement",
                   published = "1790",
                   stringsAsFactors = FALSE)
philDB$books <- rbind(philDB$books, temp)

rm(temp) #I like my code like I like my room: clean and tidy :)

temp <- data.frame(id = autoincrement(philDB$works),
                  personID = tempPersonID,
                  bookID = tempBookID1,
                  stringsAsFactors = FALSE)
philDB$works <- rbind(philDB$works, temp)

rm(temp) #I like my code like I like my room: clean and tidy :)

temp <- data.frame(id = autoincrement(philDB$works),
                   personID = tempPersonID,
                   bookID = tempBookID2,
                   stringsAsFactors = FALSE)
philDB$works <- rbind(philDB$works, temp)

rm(temp) #I like my code like I like my room: clean and tidy :)
rm(tempPersonID) #I like my code like I like my room: clean and tidy :)
rm(tempBookID1) #I like my code like I like my room: clean and tidy :)
rm(tempBookID2) #I like my code like I like my room: clean and tidy :)

# --- SECTION 2: Output information in desired format -----------------
#    Write and submit code that lists the books in alphabetical order,
#    followed by the author and the year of publishing. Format your output like:
#    "Analects" - Kongzi (220 BCE)
#    Show the result.

sel <- order(philDB$books$title)
pID <- philDB$books$id[sel]
sel <- numeric()
for (ID in pID) {
  sel <- philDB$works$personID[which(philDB$works$bookID == ID)]
  cat(sprintf("\"%s\" - %s (%s)", philDB$books$title[ID], philDB$person$name[sel], philDB$books$published[ID]))
  cat("\n")
}

rm(sel) #I like my code like I like my room: clean and tidy :)
rm(pID) #I like my code like I like my room: clean and tidy :)
rm(ID) #I like my code like I like my room: clean and tidy :)

#   The output of this is:
#   "Analects" - Kongzi (220 BCE)
#   "Being and Time" - Martin Heidegger (1927)
#   "Critique of Judgement" - Immanuel Kant (1790)
#   "Critique of Pure Reason" - Immanuel Kant (1781)
#   "Daodejing" - Laozi (530 BCE)
#   "On the Way to Language" - Martin Heidegger (1959)
#   "Zhuangzi" - Zhuangzi (300 BCE)

# END
</source>

My E-value is: 6e-39.

The NCBI protein database page link is: [https://www.ncbi.nlm.nih.gov/protein/XP_018701899.1?report=genbank&log$=protalign&blast_rank=1&RID=W4ASBWSC015 here]

My <code>MBP1_BABIN.json</code> file is:
<source lang="javascript">
[
  { "name" : "MBP1_ISAFU",
    "RefSeqID" : "XP_018701899",
    "UniProtID" : "A0A167PNW1",
    "taxonomyID" : 1081104,
    "sequence" : [
       "MVKAAPPPPP TGPGIYSATY SGIPVYEYQF GHELKEHVMR RRHDDWINAT",
       "HILKAAGFDK PARTRILERD VQKDVHEKIQ GGYGKYQGTW IPLESGEALA",
       "HRHSVYDRLR PMFEYVPGDQ SPPPAPRHAS KPKAPKVKPP LPKWGASKPK",
       "KSPAVATETH TLVPGTVPMR DDFENTDSMI VDDDTPDNIT IASASYMADE",
       "DRYDMAHVST GHRRRKREEH LHDLAEQQHA MYGDELLDYF LLSRNGGQAI",
       "VKPDPPPNFQ SDWPIDAENH TALHWACAMG DISVVRQLKR FNASISAKNA",
       "RGETPFMRSV NFTNCYERQT FSDVMKELWD TVTAQDVSGC TVIHHAAIMK",
       "NCRLYSPTCS RFYLDSILTR LDSHLSPSAL QQMLDIQDSD GNTALHLAAQ",
       "RNARKCVRSL LGRNASTDIP NNEGIRAEDL IAELNAAKKD PGPQRSSSPF",
       "APDSQRHASF RDALKDKPPK KTRKIFGSAA ATTVQSRIAP LLEEKFADLA",
       "KSYDEEWNVK DRAESEARRI LTNTQAELQA AHEQIAKLET QLEPDEVATQ",
       "VTNEANLAKH HVLSLITHQN RLHISQTADG ELSRINGDGG QDESYEERLN",
       "LARQLSHMIA EQRIAETEYV DALSMVGVGD KIDKYRRLLK RCLDPKDSES",
       "LDNNIDSVIG LMEEDRDLQN RHGPAEEPDP MDIPVGI"
       ]
  }
]
</source>

My <code>MYSPEtaxonomy.json</code> file is:
<source lang="javascript">
[
  { "ID" : 1081104,
    "species" : "Isaria fumosorosea"}
]
</source>

The result of the <code>biCode(myDB$taxonomy$species) %in% biCode(MYSPE)</code> command is:
<source lang="r">
[1] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE
</source>

The result of the <code>myDB$protein$taxonomyID %in% myDB$taxonomy$ID[(myDB$taxonomy$species == MYSPE)]</code> command is:
<source lang="r">
[1] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[14] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[27] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[40] FALSE FALSE FALSE FALSE FALSE  TRUE
</source>

; Conclusion and Outlook
As expected the only parts similar are the end. This entire unit was very fun as it helped show how we can make a rudimentary database in R. 

The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-MAT-Graphs_and_networks Graphs & Networks].

== 3.6 [http://steipe.biochemistry.utoronto.ca/abc/index.php/BIN-Databases Databases] ==
; Objective
The main objective here is to expand my knowledge of Bioinformatics databases to help introduce later databases such as the NCBI and EBI.

<div class="time-estimate">
Time estimated: 0.5 h; taken: 0.5 h; date started: 2018-10-23; date completed: 2018-10-23
</div>

; Activities
I read the introductory notes and all of the required NAR articles and issues. It's really cool, I never expected biologists to take so many leaves out of the CS book!

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/BIN-SX-Concepts Molecular Structures].

== 3.7 [http://steipe.biochemistry.utoronto.ca/abc/index.php/BIN-SX-Concepts Molecular Structures] ==
; Objective
The main objective here is to learn about how molecules are structured and their computational analysis.

<div class="time-estimate">
Time estimated: 0.5 h; taken: 1 h; date started: 2018-10-23; date completed: 2018-10-23
</div>

; Activities
This unit was tough. There is a lot I didn't understand and had to do external reading about (biologically and chemically, especially for different types of bonding)! But in the end I think I've tackled this unit well. 

The 2 main analysis methods examined were: X-ray crystallography and NMR spectroscopy (neither of these are imaging, they are methods of building consensus models). 

Other analysis methods are: homology models, electron diffraction, neutron diffraction and single molecule diffraction with X-ray lasers (atomic resolution is not trivial to implement).

It was also fun to read about certain algorithms overfitting the data creating less plausible models (perhaps ML advances may help address this in the future, such as the use of regularisers). I am excited to use ChimeraX for visualisations! I am especially excited to learn how to view stereo images. 

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/BIN-PDB Structure Databases]

== 3.8 [http://steipe.biochemistry.utoronto.ca/abc/index.php/BIN-PDB Structure Databases] ==
; Objective
The main purpose here is to learn how to understand the purpose of and learn how to navigate PDB databases (in particular the RCSB PDB). 

<div class="time-estimate">
Time estimated: 0.5 h; taken: 1 h; date started: 2018-10-23; date completed: 2018-10-23
</div>

; Activities
I read through all the nodes and navigated the RCSB PDB thoroughly. I identified the following three Saccharomyces cerevisiae Mbp1 transcription factor structures (listed by PDB ID): [http://www.rcsb.org/structure/1MB1 1MB1], [http://www.rcsb.org/structure/1L3G 1L3G] & [http://www.rcsb.org/structure/1BM8 1BM8]. The 1BM8 with all the settings as desired (coloured by hydrophobicity with liquorice style) looks like this:

[[File:Tom_1BM8.png|600px|middle]]

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-STA-Information_theory Information Theory].

== 3.9 [http://steipe.biochemistry.utoronto.ca/abc/index.php/BIN-FUNC-Concepts Biomolecular Function Concepts] ==
; Objective
Here I want to learn the main concepts behind biomolecular functions (in particular representation, annotation and prediction).

<div class="time-estimate">
Time estimated: 0.5 h; taken: 0.5 h; date started: 2018-10-23; date completed: 2018-10-23
</div>

; Activities
A short and sweet unit. Bioinformatics function does not exist. I'm happy with that. Covered all the lecture notes with some extra wiki searches as usual. 

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/RPR-FASTA FASTA].

== 3.10 [http://steipe.biochemistry.utoronto.ca/abc/index.php/BIN-ALI-Similarity Measuring Sequence Similarity] ==
; Objective
Here I want to figure out how to measure similarity of amino acid sequences.

<div class="time-estimate">
Time estimated: 0.5 h; taken: 0.5 h; date started: 2018-10-24; date completed: 2018-10-??
</div>

; Activities


; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/BIN-SYS-Concepts System models].

== 3.11 Extra learning units ==
; Objective
The main objective is to read through extra fundamental modules that are not necessary to complete the course to make sure I am fully equipped. 

<div class="time-estimate">
Time estimated: 0.5 h; taken: 0.5 h; date started: 2018-10-05; date completed: 2018-10-05
</div>

; Activities
So far the only units I have checked out are [http://steipe.biochemistry.utoronto.ca/abc/index.php/BIN-Genome-Human_genomics Human Genomics], [http://steipe.biochemistry.utoronto.ca/abc/index.php/BIN-ALI-Dotplot Dotplot], [http://steipe.biochemistry.utoronto.ca/abc/index.php/BIN-ALI-Internal_repeats Internal Repeats], [http://steipe.biochemistry.utoronto.ca/abc/index.php/BIN-PHYLO-Selective_pressure Phylogenetic Analysis], [http://steipe.biochemistry.utoronto.ca/abc/index.php/BIN-PHYLO-Conservation_scores Conservation Scores], [http://steipe.biochemistry.utoronto.ca/abc/index.php/BIN-SX-Motifs Structure Motifs], [http://steipe.biochemistry.utoronto.ca/abc/index.php/BIN-SX-MD Molecular Dynamics], [http://steipe.biochemistry.utoronto.ca/abc/index.php/BIN-SX-Ab_initio_prediction Prediction] and [http://steipe.biochemistry.utoronto.ca/abc/index.php/BIN-Genome-NGS_bioinformatics Next Generation Sequencing], which are still stubs.

; Conclusion and Outlook
Although this was not required, it was definitely worthwhile!

= 4. Statistics =
This section will be focussing on learning units that concern mathematical statistical theory and it implications in bioinformatics.

== 4.1 [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-STA-Probability Probability] ==
; Objective
As a student in Machine Learning, the aim of this course is to refresh my knowledge on basic probability theory.

<div class="time-estimate">
Time estimated:0.5 h; taken: 0.5 h; date started: 2018-10-05; date completed: 2018-10-05
</div>

; Activities
This was fun. The touch of humour of Prof. Steipe is definitely appreciated! I liked the examples chosen too. Worked through the entire unit without too much trouble.

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-STA-Probability_distribution Probability Distribution].

== 4.2 [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-STA-Probability_distribution Probability Distribution] ==
; Objective
As a student in Machine Learning and having already covered KL divergence, the aim of this course is to refresh my knowledge on probability distributions.

<div class="time-estimate">
Time estimated: 1 h; taken: 1 h; date started: 2018-10-05; date completed: 2018-10-05
</div>

; Activities
I went through the R code thoroughly. There was actually a lot of very useful material that I had certainly forgotten or not previously considered (taken for granted). And as always I love all those colourful R graphs. 

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-STA-Significance Significance].

== 4.3 [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-STA-Significance Significance] ==
; Objective
Since I did my bachelor thesis on granger causality relations of blockchain financial time series hopefully this chapter should be more of a consolidation than tackling new material. The main aim is to ensure I fully understand how to use and manipulate p-values.

<div class="time-estimate">
Time estimated: 1 h; taken: 1 h; date started: 2018-10-05; date completed: 2018-10-06
</div>

; Activities
I went through the R code thoroughly and read the recommended papers. The p value I got using <code> p = (Nobs + 1) / (N + 1) </code> was <code> 0.03297888 </code>. This suggests statistically significant evidence against the null hypothesis (which in this case is that positively charged residues are not closer to negatively charged residues). The code also asks to mark down the sum of all the "chs" in binary hex format (I assume for Prof. Steipe to verify). Mine is: <code> 00 00 00 00 8c 9a e1 40 </code>.

The [https://www.ncbi.nlm.nih.gov/pubmed/26961635 first paper] reviewed basically describes the tears of the American Statistical Association (ASA) which are concerned that papers only get published when p-values indicated statistical significance while papers showing no statistical significance do not get published. This yields a falsely robust methodology that the ASA are hoping to reverse.

The [https://www.ncbi.nlm.nih.gov/pubmed/21878926 second paper] reviewed criticises neuroscience publications finding that around 50% of them use the incorrect procedure for evaluating statistical significance between two events, say A and B. The main error being that often A is shown to be statistically significant and control event B is said to be statistically significant but the difference between A and B is not examined to see if it itself is statistically significant. This paper also criticises the Cliff effect whereby people are overly swayed once the artificial boundary of $p < 0.05$ is crossed. It however concludes than in many cases this lack of statistic rigour would not however change the conclusions of the papers.

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-Genetic_code Genetic Code].

== 4.4 [http://steipe.biochemistry.utoronto.ca/abc/index.php/FND-STA-Information_theory Information Theory] ==
; Objective
Since I am already quite familiar with information theory hopefully this should be a refresher.

<div class="time-estimate">
Time estimated: 0.5 h; taken: 0.5 h; date started: 2018-10-23; date completed: 2018-10-23
</div>

; Activities
I went through the notes thoroughly and made sure I knew everything. This was not too challenging for me given my background.

; Conclusion and Outlook
The next unit that I plan to take is [http://steipe.biochemistry.utoronto.ca/abc/index.php/BIN-FUNC-Concepts Biomolecular Function Concepts].

=Notes and References=
<references />

{{Vspace}}

<div class="note">
Note: This page is still in development. 
</div>


{{CC-BY}}

[[Category:BCH441-2018 Journal]]
